### PR TITLE
Codestyle + Symbolic Random for all(hope all) Distributions

### PR DESCRIPTION
**Pull request by ferrine**
better codestyle

### CODE REVIEW

File: pymc3/distributions/discrete.py [hunk 1]
```diff
@@ -295,7 +296,7 @@ class NegativeBinomial(Discrete):
     def __init__(self, mu, alpha, *args, **kwargs):
         super(NegativeBinomial, self).__init__(*args, **kwargs)
         self.mu = mu = tt.as_tensor_variable(mu)
-        self.alpha = alpha = tt.as_tensor_variable(alpha)
```
Comment (twiecki) [inline pymc3/distributions/discrete.py]: some of these had a point that below where we construct the self.mean we want that to be a theano expression too. It's only safe to remove this if we use self.alpha below too. And that wasn't the case always.

**Comment by ferrine** [comment]
@twiecki While I was working on symbolic random you did a great point so that I decided not to do separate PRs that are both refactoring

**Comment by twiecki** [comment]
Nice, like the refactoring of the random number generation.

**Comment by fonnesbeck** [comment]
Is there a reason for changing the method name from `random` to `rng`? Seems like it would cause unnecessary breakage.

File: pymc3/distributions/continuous.py [hunk 2]
```diff
@@ -203,26 +201,22 @@ class Normal(Continuous):
         Precision (tau > 0).
     """
 
+    rng = stats.norm.rvs
```
Comment (fonnesbeck) [inline pymc3/distributions/continuous.py:201]: We should wrap the SciPy's `rvs` function to use PyMC argument names. SciPy uses generic `loc` and `scale` names, which is confusing. This has been done for some of the distributions I see, but not all.

**Comment by ferrine** [comment]
@fonnesbeck `rgn` is function that accepts size and not dist_shape, `random` utilizes it and reshapes according to passed dist_shape in generate_samples wrapper over generators

File: pymc3/distributions/continuous.py (hunk 2 above)
Comment (ferrine) [reply pymc3/distributions/continuous.py:201]: I pass arguments to self.parent ordered dict and then don't care about their names user should not too.

**Comment by fonnesbeck** [comment]
Ah, I see it now in the `Distribution` parent class. Thanks for the clarification.

File: pymc3/distributions/distribution.py [hunk 3]
```diff
@@ -375,110 +392,94 @@ def generate_samples(generator, *args, **kwargs):
...
-
-    def logp(self, value):
-        return bound(self.dist.logp(value),
-                     value >= self.lower, value <= self.upper)
+class _BoundedIndicator(object):
+    pass
 
 
-class Bound(object):
+def Bounded(distribution, lower=-np.inf, upper=np.inf, transform='infer'):
```
Comment (twiecki) [inline pymc3/distributions/distribution.py:401]: What is the purpose of this refactor?

Comment (ferrine) [reply pymc3/distributions/distribution.py:401]: Now bounded distributions are not subclasses of `Distribution`, I want to add it for consistent `isinstance` check

Comment (twiecki) [reply pymc3/distributions/distribution.py:401]: Seems quite costly in terms of added code complexity. I suppose `isinstance(Bound(), Distribution)` returns True?

Comment (ferrine) [reply pymc3/distributions/distribution.py:401]: No it is `object` instance
https://github.com/pymc-devs/pymc3/blob/master/pymc3/distributions/distribution.py#L453

Comment (twiecki) [reply pymc3/distributions/distribution.py:401]: What doe we need the isinstance check for? You do agree, that we have to do some ugly stuff below to get there, right?

File: pymc3/distributions/distribution.py [hunk 4]
```diff
@@ -375,110 +392,94 @@ def generate_samples(generator, *args, **kwargs):
...
+     par = boundedNormal(mu=0.0, sd=1.0, testval=1.0)
+     """
+    from pymc3.distributions.transforms import (
+        interval, lowerbound, upperbound
+    )
+    import inspect
+    if issubclass(distribution, _BoundedIndicator):
+        raise ValueError('Cannot bound already bounded Distribution')
+
+    class Dummy(distribution, _BoundedIndicator):
```
Comment (fonnesbeck) [inline pymc3/distributions/distribution.py]: We should avoid multiple inheritance if possible.

**Comment by ferrine** [comment]
Finally, I found the reason of failing test:
```
samples[samples.size-np.isnan(samples).sum():][:select.size]=select # work
samples[np.isnan(samples)][:select.size]=select # does not work
```

File: pymc3/distributions/distribution.py (hunk 4 above)
Comment (ferrine) [reply pymc3/distributions/distribution.py]: How do I prevent user from bounding bounded distribution?

File: pymc3/distributions/distribution.py (hunk 3 above)
Comment (ferrine) [reply pymc3/distributions/distribution.py:401]: I found this problem here https://github.com/ferrine/gelato/blob/master/gelato/spec.py#L50

**Comment by ferrine** [comment]
Some distributions were impossible to implement in static way:( I've left NotImplemented exception for that case

**Comment by twiecki** [comment]
@ferrine is this still WIP or ready for review?

**Comment by ferrine** [comment]
Symbolic random is still left

**Comment by ferrine** [comment]
I've made this simple code work, but it is not picklable, do you have any idea how to fix it?

```python
class SymbRand(object):
    def __init__(self, rv):
        par = rv.distribution.parents
        self.rv = rv
        self.args = par.keys()
        self.vals = par.values()
        self.rng = rv.distribution.rng

    def __call__(self, size=None):
        itypes = [tt.lscalar] + [v.type for v in self.vals]
        if size is not None:
            otype = tt.TensorType(
                self.rv.dtype,
                (False, ) + self.rv.broadcastable
            )
        else:
            otype = self.rv.type
        if size is None:
            size = np.asarray([], 'int64')
        ndim = self.rv.ndim+size.size
        size = tt.as_tensor(size)
            
        @theano.as_op(
             itypes, [otype]
        )
        def tmp_op(*args):
            kwargs = dict(zip(self.args, args[1:]))
            size = args[0]
            sample = self.rng(size=(size, ), **kwargs)
            return sample
        return tmp_op(size.prod()*self.rv.shape.prod(), *self.vals).reshape(tt.concatenate([size, self.rv.shape]),  ndim=ndim)

with pm.Model():
    n = pm.Normal('n')
theano.clone(SymbRand(n)([10, 2, 1]), {n:tt.constant(1., dtype='float64')}).eval()
```

**Comment by fonnesbeck** [comment]
Might be because you have a function as an attribute (`rv.distribution.rng`).

**Comment by ferrine** [comment]
It is because I define Op inside __call__

**Comment by ferrine** [comment]
Don't think I'll continue working on this

**Pull request closed without merging**